mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,277 @@
+
+    #[inline]
+    pub(crate) fn sender_peer(&self) -> &str { &self.data.sender_peer }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[cfg_attr(any(test, target_arch = "wasm32"), derive(PartialEq))]
+struct HealthcheckData {
+    sender_peer: String,
+    sender_public_key: Vec<u8>,
+    target_peer: String,
+    expires_at: i64,
+    is_a_reply: bool,
+}
+
+impl HealthcheckData {
+    #[inline]
+    fn encode(&self) -> Result<Vec<u8>, rmp_serde::encode::Error> { encode_message(self) }
+}
+
+#[inline]
+pub fn peer_healthcheck_topic(peer_id: &PeerId) -> String {
+    pub_sub_topic(PEER_HEALTHCHECK_PREFIX, &peer_id.to_string())
+}
+
+#[derive(Deserialize)]
+pub struct RequestPayload {
+    peer_id: String,
+}
+
+#[derive(Debug, Display, Serialize, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum HealthcheckRpcError {
+    InvalidPeerAddress { reason: String },
+    MessageGenerationFailed { reason: String },
+    MessageEncodingFailed { reason: String },
+}
+
+impl HttpStatusCode for HealthcheckRpcError {
+    fn status_code(&self) -> common::StatusCode {
+        match self {
+            HealthcheckRpcError::InvalidPeerAddress { .. } => StatusCode::BAD_REQUEST,
+            HealthcheckRpcError::MessageGenerationFailed { .. } | HealthcheckRpcError::MessageEncodingFailed { .. } => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+pub async fn peer_connection_healthcheck_rpc(
+    ctx: MmArc,
+    req: RequestPayload,
+) -> Result<bool, MmError<HealthcheckRpcError>> {
+    /// When things go awry, we want records to clear themselves to keep the memory clean of unused data.
+    /// This is unrelated to the timeout logic.
+    const ADDRESS_RECORD_EXPIRATION: Duration = Duration::from_secs(60);
+
+    let target_peer_id = PeerId::from_str(&req.peer_id)
+        .map_err(|e| HealthcheckRpcError::InvalidPeerAddress { reason: e.to_string() })?;
+
+    let p2p_ctx = P2PContext::fetch_from_mm_arc(&ctx);
+
+    if target_peer_id == p2p_ctx.peer_id() {
+        // That's us, so return true.
+        return Ok(true);
+    }
+
+    let message =
+        HealthcheckMessage::generate_message(&ctx, target_peer_id, false, ctx.healthcheck_config.message_expiration)
+            .map_err(|reason| HealthcheckRpcError::MessageGenerationFailed { reason })?;
+
+    let encoded_message = message
+        .encode()
+        .map_err(|e| HealthcheckRpcError::MessageEncodingFailed { reason: e.to_string() })?;
+
+    let (tx, rx): (Sender<()>, Receiver<()>) = oneshot::channel();
+
+    let mut book = ctx.healthcheck_response_handler.lock().await;
+    book.clear_expired_entries();
+    book.insert(target_peer_id.to_string(), tx, ADDRESS_RECORD_EXPIRATION);
+    drop(book);
+
+    broadcast_p2p_msg(&ctx, peer_healthcheck_topic(&target_peer_id), encoded_message, None);
+
+    let timeout_duration = Duration::from_millis(ctx.healthcheck_config.timeout_ms);
```

Comment:
Also this timeout might be confusing to set as it's not really visible what it stands for at a first glance.

We know won't get back a response if the message was expired before it reaches the node in question. So we might just double (or triple, for safety) that and that should be the maximum time we wait for hoping that there is a response coming.

Put simpler, let's remove this config parameter and timeout after `2 * ctx.healthcheck_config.message_expiration`.

Reply:
This logic doesn't consider message expiration at all. Message expiration and RPC timeout serve entirely different purposes and I don't think we should make a connection between these two. I might want to set my RPC timeout to 5 seconds while sending a message that remains valid for 30 seconds. Nodes can send health check messages (as a reply to the other nodes) even without invoking the RPC method.

Reply:
Q: then, does message expiration has any meaning when requesting healthcheck (not responding)?

Reply:
See https://github.com/KomodoPlatform/komodo-defi-proxy/pull/25#discussion_r1772650890.